sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesAccessCorrectField.cs

```diff
@@ -0,0 +1,258 @@
+    [Rule(DiagnosticId)]
+    public sealed class PropertiesAccessCorrectField : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4275";
+        private const string MessageFormat = "Refactor this {0} so that it actually refers to the field '{1}'";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            // We want to check the fields read and assigned in all properties in this class
+            // so this is a symbol-level rule (also means the callback is called only once
+            // for partial classes)
+            context.RegisterSymbolAction(CheckType, SymbolKind.NamedType);
+        }
+
+        private void CheckType(SymbolAnalysisContext context)
+        {
+            var symbol = (INamedTypeSymbol)context.Symbol;
+            if (symbol.TypeKind != TypeKind.Class &&
+                symbol.TypeKind != TypeKind.Structure)
+            {
+                return;
+            }
+
+            var fields = symbol.GetMembers().Where(m => m.Kind == SymbolKind.Field).OfType<IFieldSymbol>();
+            if (!fields.Any())
+            {
+                return;
+            }
+
+            var properties = GetExplictlyDeclaredProperties(symbol);
+            if (!properties.Any())
+            {
+                return;
+            }
+
+            var propertyToFieldMatcher = new PropertyToFieldMatcher(fields);
+            var allPropertyData = CollectPropertyData(properties, context.Compilation);
+
+            // Check that if there is a single matching field name it is used by the property
+            foreach (var data in allPropertyData)
+            {
+                var expectedField = propertyToFieldMatcher.GetSingleMatchingFieldOrNull(data.PropertySymbol);
+                if (expectedField != null)
+                {
+                    CheckExpectedFieldIsUsed(expectedField, data.FieldUpdated, context);
+                    CheckExpectedFieldIsUsed(expectedField, data.FieldReturned, context);
+                }
+            }
+        }
+
+        private static IEnumerable<IPropertySymbol> GetExplictlyDeclaredProperties(INamedTypeSymbol symbol) =>
+            symbol.GetMembers()
+                .Where(m => m.Kind == SymbolKind.Property)
+                .OfType<IPropertySymbol>()
+                .Where(p => !p.IsImplicitlyDeclared);
+
+        private static void CheckExpectedFieldIsUsed(IFieldSymbol expectedField, FieldData? actualField, SymbolAnalysisContext context)
+        {
+            if (actualField.HasValue && actualField.Value.Field != expectedField)
+            {
+                context.ReportDiagnosticWhenActive(Diagnostic.Create(
```

Comment:
this could be reduced on less lines, even probably only one, if you extract the ternary operator to a variable.

Reply:
I would find that less readable. Does reducing the number of lines matter?

Reply:
It is for style consistency, we don't put each argument on a separate line.